posts/01-effective-rust-canisters.html.pm

```diff
@@ -442,6 +442,355 @@ type CreateEntityResult = variant {
+  You should apply the same technique for all types deriving serde's ◊code{Serialize} and ◊code{Deserialize} traits, not just for types you encode as Candid.
+  A ◊a[#:href "https://github.com/dfinity/ic/commit/1b98a5d984176b1c948d0cb92227d88ad5ee8044"]{similar change} boosted the ICP ledger archive upgrades (the canister uses ◊a[#:href "https://cbor.io"]{CBOR} for state serialization).
+}
+
+◊advice["avoid-copies"]{Avoid copying large values.}
+
+◊p{
+  Experience shows that canisters spend a lot of their instructions copying bytes◊sidenote["sn-bulk-ops"]{
+    Spending a lot of time in ◊code{memcpy} and ◊code{memset} is a common trait of many WebAssembly programs.
+    That observation led to the ◊a[#:href "https://github.com/WebAssembly/bulk-memory-operations/blob/dcaa1b6791401c29b67e8cd7929ec80949f1f849/proposals/bulk-memory-operations/Overview.md"]{bulk memory operations} proposal included in the ◊a[#:href "https://webassembly.github.io/spec/core/appendix/changes.html?highlight=proposals#bulk-memory-and-table-instructions"]{WebAssembly 2.0 release}.}.
+  Reducing the number of unnecessary copies often affects cycle consumption.
+}
+◊p{
+  Let us imagine that we work on a canister that serves a single dynamic asset.
+}
+◊source-code["rust"]{
+thread_local!{
+    static ASSET: RefCell<Vec<u8>> = RefCell::new(init_asset());
+}
+
+#[derive(CandidType, Deserialize)]
+struct HttpResponse {
+    status_code: u16,
+    headers: Vec<(String, String)>,
+    #[serde(with = "serde_bytes")]
+    body: Vec<u8>,
+}
+
+#[query]
+fn http_request(_request: HttpRequest) -> HttpResponse {
+    ◊em{// ◊b{NOTE}: we are making a full copy of the asset.}
+    let body = ASSET.with(|cell| cell.borrow().clone());
+
+    HttpResponse {
+        status_code: 200,
+        headers: vec![("Content-Length".to_string(), body.len().to_string())],
+        body
+    }
+}
+}
+◊p{
+  The ◊code{http_request} endpoint makes a deep copy of the asset for every request.
+  This copy is unnecessary because the CDK encodes the response into the reply buffer as soon as the endpoint returns.
+  There is no need for the encoder to own the body.
+  Unfortunately, the current macro API makes it unnecessarily hard to eliminate copies: the type of reply must have ◊code{'static} lifetime.
+  There are a few ways to work around this issue.
+}
+◊p{
+  One solution is to wrap the asset body into a ◊a[#:href "https://doc.rust-lang.org/std/sync/struct.Arc.html"]{reference-counting smart pointer}.
+}
+◊figure{
+◊marginnote["mn-rc-bytes"]{
+  Using a reference-counting pointer for large values.
+  Note that the type of the ◊code{ASSET} variable has to change: all copies of the data must be behind the smart pointer.
+}
+◊source-code["rust"]{
+thread_local!{
+    static ASSET: RefCell<RcBytes> = RefCell::new(init_asset());
+}
+
+struct RcBytes(Arc<serde_bytes::ByteBuf>);
```

Comment:
Do you actually need `Arc` here? I'd expect `Rc` to be enough since we're in a single thread mode anyway.

Reply:
I started with Rc (which I used in the certified assets), but then I realized that Rc might be a PITA in async code (Rc is not Send). I measured Arc and Rc; there is no difference in cycle consumption in the Wasm context. I'd say there is no good reason to prefer Rc to Arc in Wasm code.

Reply:
> I started with Rc (which I used in the certified assets), but then I realized that Rc might be a PITA in async code (Rc is not Send)

That's a good point about the async code. I was mostly asking out of curiosity, thanks for the answer!